Add finalizers and proper thread scanning to SDAlloc

schveiguy opened the pull request:
This adds the calling of finalizers on collection *FOR DENSE BLOCKS ONLY*. I intend to add them for sparse blocks in a future update.

For integration with druntime, this must use my branch: https://github.com/schveiguy/dmd/tree/addsdcgc which adds some necessary hooks.

This also fixes an issue with `thread_scanAll`. In the current druntime, `thread_scanAll` is a function which accepts a D delegate using start and end pointer. However, the ABI between SDC and DMD is not defined. So this has to be done via extern(C) functions. This is why we are doing things in an even more convoluted way (with `__sd_thread_scanAll`).

deadalnix commented:
This is doing many things at once. At least 3:
1. Add finalization for dense allocations.
2. Fix the issue where druntime and the GC do not communicate the range to scan for the stack properly.
3. Address performance problems in __sd_gc_druntime_qalloc .

Ultimately, it'd be better if this was 3 PR instead of one. Any comment on any of these points will block the other 2, which isn't what we want. In addition some of this code could go into master, while the  rest still rely ion a WIP on druntime side and likely shouldn't be merged in master right away.

But first and foremost, let's make sure we stick with the style of the project. `tool/autoformat` can help greatly there.

deadalnix commented on sdlib/d/gc/capi.d:
```diff
@@ -7,6 +7,7 @@ extern(C):
 /**
  * Standard C allocating functions.
  */
+/+
```
I don't understand why this is here. This is already commented in sdc's branch. Is github doing something retarded?

schveiguy replied:
I can only say, I think so. I agree it does look like this is already there.

schveiguy commented:
OK, I rebased on master, and rebased on druntimegc. But your druntimegc does not include the extern(C) changes to dmd/thread.d, so it looks like I'm redoing it here. But really, this is from master. I'm not sure what to do here.

I'll at least work on getting the other parts split out.

schveiguy commented:
> 3\. Address performance problems in __sd_gc_druntime_qalloc .

Hm... I don't think this is right. What I did was add back in the code that was *bad*, but just commented out. We don't know if we may need it later...

schveiguy commented:
OK, I think this is ready to be looked at again.

schveiguy commented:
Oh, I can probably do the sparse blocks too. Will work on that.

schveiguy commented:
Alright, the finalizer support is all there. I still don't know why 32-byte blocks don't get collected correctly. Am investigating. But it has nothing to do (I think) with the finalizer changes.

deadalnix commented on sdlib/d/finalizer.d:
```diff
@@ -0,0 +1,9 @@
+// SDC implementation of finalizer for the GC.
+
+extern(C)
+void __sd_destroyBlockCtx(void* ptr, size_t usedSpace, void* finalizer) {
```
`__sd_run_finalizer` ?

deadalnix commented on sdlib/d/gc/page.d:
```diff
@@ -706,6 +709,7 @@ private:
 
 				import d.gc.slab;
 				auto nslots = binInfos[sc].nslots;
+				auto ssize = binInfos[sc].slotSize;
```
If the block has a finalizer, it should have its actual size in its metadata.

schveiguy replied:
When looking up the metadata, I have an address and an index. The calculation to get the actual block pointer is:

```d
auto ptr = e.address + idx * ssize;
```

Is there a better way to do this? How do I get at the metadata before knowing the slot boundaries?

deadalnix replied:
I see. but you only need that when destroying, right? So why is this here?

schveiguy replied:
Moved until later, where there are definitely going to be finalizers called. I was thinking it would save some calculation while looping over the bits, since the slot size does not change, and I thought this was instead of doing it per finalizer. But I can do it per nimble instead, so it will be less impactful.

deadalnix commented on sdlib/d/gc/page.d:
```diff
@@ -725,6 +729,42 @@ private:
 					auto oldOccupancy = e.slabData.rawContent[i];
 					auto newOccupancy = oldOccupancy & bmp[i];
 
+					if (ec.supportsMetadata) {
```
If you put that block at the end of the loop, you can early continue rather than nest like migrating birds.

schveiguy replied:
I need to do this before setting the bits in the slab, right? I have rewritten it, but I have to copy that line.

deadalnix commented on sdlib/d/gc/page.d:
```diff
@@ -725,6 +729,42 @@ private:
+								}
+
+								// clear the bit
+								toRemove ^= 1UL << bit;
```
`toRemove &= (toRemove - 1)`

schveiguy commented:
updated. Still some remaining questions.

schveiguy commented:
rebased, I need to test for performance and correctness.

deadalnix commented on sdlib/d/gc/bitmap.d:
```diff
@@ -48,6 +48,13 @@ public:
 		return (n & 0x01) != 0;
 	}
 
+	ulong rawNimbleAtomic(uint nimble) shared {
```
Type qualifier are not well supported in SDC (they never where) so putting them in the code is mostly aspirational :P

deadalnix replied:
Nevertheless, const shared is a thing and as long as the name is explicit that should be fine.

schveiguy replied:
OK, so make an accessor that returns const shared slice? I can do that.

deadalnix commented on sdlib/d/gc/tcache.d:
```diff
@@ -8,11 +8,16 @@ import d.gc.slab;
 import d.gc.spec;
 import d.gc.tbin;
 import d.gc.util;
+import d.gc.extent;
```
Remove

deadalnix commented on sdlib/d/gc/tcache.d:
```diff
@@ -8,11 +8,16 @@ import d.gc.slab;
 
 import sdc.intrinsics;
 
+import core.stdc.stdio;
```
Remove.

deadalnix commented on sdlib/dmd/gc.d:
```diff
@@ -3,6 +3,7 @@ module dmd.gc;
 import d.gc.tcache;
 import d.gc.spec;
 import d.gc.slab;
+import d.gc.emap;
```
You should not be snooping in the emap without going through the thread cache.

deadalnix commented on sdlib/d/gc/page.d:
```diff
@@ -11,6 +11,8 @@ import d.gc.util;
 
 import sdc.intrinsics;
 
+extern(C) void __sd_run_finalizer(void* ptr, size_t usedSpace, void* finalizer);
```
You should import that declaration instead of duplicating it all over the place.

schveiguy replied:
I had already declared it in d/finalizer.d, but I didn't know how the linking worked. I thought if I imported that file, it would include the implementation, not realizing the object is not built for it.

deadalnix commented on sdlib/d/gc/page.d:
```diff
@@ -728,6 +731,44 @@ private:
 					count += popCount(oldOccupancy ^ newOccupancy);
 
+					if (!ec.supportsMetadata) {
+						e.slabData.rawContent[i] = newOccupancy;
```
This is duplicated, so really not the ideal setup. generally, this code has too many nesting levels, it clearly needs a refactoring. incidentally, part of the refactoring might also apply to sparse slabs.

schveiguy replied:
Switched to a scope(exit), which works nicely, and saves another indent.

deadalnix commented on sdlib/d/gc/page.d:
```diff
@@ -728,6 +731,44 @@ private:
+					}
+
+					auto toRemove = (oldOccupancy ^ newOccupancy)
+						& e.slabMetadataFlags.rawNimbleAtomic(cast(uint) i);
```
Make i a uint?

schveiguy replied:
With the new API of returning the reference to the raw static array, this is moot.

deadalnix commented on sdlib/dmd/gc.d:
```diff
@@ -71,17 +72,46 @@ void __sd_gc_druntime_qalloc(BlkInfo* result, size_t size, uint bits,
-	if (appendable) {
+	if (appendable || finalizer !is null) {
+		import core.stdc.stdio;
+		//printf("In sdc qalloc, with finalizer or appendable: %p\n", finalizer);
```
Remove.

deadalnix commented on sdlib/dmd/gc.d:
```diff
@@ -71,17 +72,46 @@ void __sd_gc_druntime_qalloc(BlkInfo* result, size_t size, uint bits,
-			.allocAppendable(size, hasPointers, hasPointers,
-			                 cast(Finalizer) finalizer);
+			.allocAppendable(size, hasPointers, hasPointers, finalizer);
+		//printf("allocated address %p\n", result.base);
```
dito

deadalnix commented on sdlib/dmd/gc.d:
```diff
@@ -71,17 +72,46 @@ void __sd_gc_druntime_qalloc(BlkInfo* result, size_t size, uint bits,
 	result.size = size;
-	result.attr = bits & (BlkAttr.APPENDABLE | BlkAttr.NO_SCAN);
+
+	// note, we may not need this code, and probably shouldn't use it.
+	/*if (result.base) {
+		if (appendable) {
+			// figure out the capacity, set it to max, and then use that size
+			// for the caller.
+			auto cap = threadCache.getCapacity(result.base[0 .. size]);
+			if (cap == 0) {
+				result.size = size;
+			} else {
+				assert(threadCache.extend(result.base[0 .. size], cap - size));
+				result.size = cap;
+			}
+		} else {
+			// no good mechanism to look this up, so wing it
+			auto pd = threadCache.getPageDescriptor(result.base);
+			if (pd.isSlab()) {
+				auto si = SlabAllocInfo(pd, result.base);
+				result.size = si.usedCapacity;
+			} else {
+				auto e = pd.extent;
+				result.size = e.usedCapacity;
+			}
+		}
+	}*/
```
dito

deadalnix commented on sdlib/dmd/gc.d:
```diff
@@ -110,3 +140,59 @@ BlkInfo __sd_gc_druntime_allocInfo(void* ptr) {
+		}
+	}
+}
+
+bool __sd_setArrayUsed(void* ptr, size_t pdData, size_t newUsed,
+                       size_t existingUsed, bool atomic) {
+	auto pd = PageDescriptor(pdData);
+	auto e = pd.extent;
+	if (e) {
+		if (pd.isSlab()) {
+			auto si = SlabAllocInfo(pd, ptr);
+			if (existingUsed == -1 || si.usedCapacity == existingUsed) {
+				si.setUsedCapacity(newUsed);
+				return true;
+			}
+		} else {
+			if (existingUsed == -1 || e.usedCapacity == existingUsed) {
+				e.setUsedCapacity(newUsed);
+				return true;
+			}
+		}
+	}
+
+	return false;
+}
+
+size_t __sd_getArrayUsed(void* ptr, size_t pdData, bool atomic) {
+	auto pd = PageDescriptor(pdData);
+	auto e = pd.extent;
+	if (e) {
+		if (pd.isSlab()) {
+			auto si = SlabAllocInfo(pd, ptr);
+			return si.usedCapacity;
+		} else {
+			return e.usedCapacity;
+		}
+	}
+
+	return 0;
+}
```
This does not look like it is necessary to destroy anything.

schveiguy replied:
It's for the array runtime stuff, which can be a separate PR. Removed.

deadalnix commented on sdlib/d/gc/page.d:
```diff
@@ -784,21 +825,63 @@ private:
+					// All set bits in hasMeta have metadata.
+					auto baseidx = i * 64;
+					while (hasMeta != 0) {
+						uint bit = countTrailingZeros(hasMeta);
```
This is called index all over the codebase.

deadalnix commented on sdlib/d/gc/page.d:
```diff
@@ -784,21 +825,63 @@ private:
+					auto baseidx = i * 64;
+					while (hasMeta != 0) {
+						uint bit = countTrailingZeros(hasMeta);
+						// NOTE: this copies techniques/code from
+						// SlabAllocInfo, but that code starts with a
+						// pointer and gives us info we already have.
+						// So we don't want to do that work again.
```
It might be worth thinking about refactorings here.

schveiguy replied:
You mean a refactor in slab.d? I need this specific calculation abstracted out I guess.

schveiguy replied:
Removed.

deadalnix commented on sdlib/d/gc/tcache.d:
```diff
@@ -8,11 +8,16 @@ import d.gc.slab;
+
 ThreadCache threadCache;
 
+extern(C) void __sd_run_finalizer(void* ptr, size_t usedSpace, void* finalizer);
```
Once again, this declaration shouldn't be duplicated all over the codebase. Where is the natural place to declare it?

schveiguy commented:
Most things resolved. Just the one question of refactoring.

schveiguy commented:
Just realizing, this should have tests. Where do you put library tests?

deadalnix reviewed (commented):
Now we'll need to have a test for this. This is why i advised to only do large as the first iteration. It'll be easy to see in subsequent patches that test cases do match the added features.

See `test/valid/test0176.d` for an example of such a test.

schveiguy commented:
Test added, should be ready to go.

schveiguy commented:
looks like the CI is trying to install some package that doesn't exist?

deadalnix commented on sdlib/d/gc/tcache.d:
```diff
@@ -103,6 +103,7 @@ public:
 	}
 
 	void destroy(void* ptr) {
+		import d.finalizer;
```
Move it where it is used.

schveiguy commented on test/valid/test0197.d:
```diff
@@ -0,0 +1,46 @@
+	}
+}
+
+void killstack() {
```
I removed the function, I added it defensively, as this is typically what I do to ensure an item is not on the stack. We'll see if it works.
